tests/components/downloader/test_services.py

```diff
@@ -0,0 +1,199 @@
+"""Test downloader action."""
+
+from http import HTTPStatus
+from unittest.mock import mock_open, patch
+
+import pytest
+from requests_mock import Mocker
+
+from homeassistant.components.downloader.const import (
+    ATTR_AUTH_PASSWORD,
+    ATTR_AUTH_TYPE,
+    ATTR_AUTH_USERNAME,
+    ATTR_SUBDIR,
+    ATTR_URL,
+    DOMAIN,
+    SERVICE_DOWNLOAD_FILE,
+)
+from homeassistant.const import HTTP_BASIC_AUTHENTICATION, HTTP_DIGEST_AUTHENTICATION
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ServiceValidationError
+
+
+def digest_challenge_matcher(request):
+    """Match requests with no auth header."""
+    return "Authorization" not in request.headers
+
+
+def digest_response_matcher(request):
+    """Match requests with an auth header."""
+    return "Authorization" in request.headers
+
+
+@pytest.mark.usefixtures("loaded_config_entry")
+async def test_service_call(hass: HomeAssistant, requests_mock: Mocker) -> None:
+    """Test downloader action."""
+
+    TEST_URL = "http://example.com/resource"
+
+    requests_mock.get(
+        TEST_URL,
+        additional_matcher=digest_response_matcher,
+        status_code=HTTPStatus.OK,
+    )
+
+    with patch("builtins.open", new=mock_open(), create=True):
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_DOWNLOAD_FILE,
+            {
+                ATTR_URL: TEST_URL,
+                ATTR_AUTH_TYPE: HTTP_BASIC_AUTHENTICATION,
+                ATTR_AUTH_USERNAME: "test",
+                ATTR_AUTH_PASSWORD: "test",
+            },
+            blocking=True,
+        )
+
+    assert requests_mock.called
+
+
+@pytest.mark.usefixtures("loaded_config_entry")
+async def test_invalid_subdir(hass: HomeAssistant, requests_mock: Mocker) -> None:
```

Comment:
The validation occurs before the request is made, so you shouldn't need to use the `requests_mock` fixture nor patch the `buildins.open`
You could also use parametrize here to have a single test with different error code.

```suggestion
@pytests.mark.parametrize(("extra_args", "error_code"), [({ATTR_SUBDIR: "~"}, "invalid_subdir"), ({ATTR_AUTH_TYPE: HTTP_BASIC_AUTHENTICATION}, "missing_credentials"), ({ATTR_AUTH_TYPE: HTTP_DIGEST_AUTHENTICATION}, "missing_credentials")])
@pytest.mark.usefixtures("loaded_config_entry")
async def test_invalid_subdir(hass: HomeAssistant, extra_args: dict[str, Any], error_code: str) -> None:
```